### PR TITLE
ENH: Allow keep='all' for nlargest/nsmallest

### LINKED ISSUE #16818: Make parameter keep=False keep duplicates for nlargest/nsmallest
#### Code Sample, a copy-pastable example if possible

```python
>>> s = pd.Series([10,9,8,7,7,7,6])
>>> s.nlargest(4)
0    10
1     9
2     8
3     7
dtype: int64
```
#### Problem description
The docstrings list `False` as one of the possible argument values for `keep`.  pandas raises a `ValueError` when attempting to use this parameter. 

#### Expected Output
It would be nice to have nlargest work like this.

```python
>>> s.nlargest(4, keep=False)
0    10
1     9
2     8
3     7
4     7
5     7
```
#### Output of ``pd.show_versions()``

<details>
# Paste the output here pd.show_versions() here
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Darwin
OS-release: 15.6.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.2
pytest: 3.0.7
pip: 9.0.1
setuptools: 35.0.2
Cython: 0.25.2
numpy: 1.13.0
scipy: 0.19.0
xarray: None
IPython: 6.0.0
sphinx: 1.5.5
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: 1.2.0
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.7
xlrd: 1.0.0
xlwt: 1.2.0
xlsxwriter: 0.9.6
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.999999999
sqlalchemy: 1.1.9
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: 0.3.0.post
</details>

### PR DESCRIPTION
**Pull request by gfyoung**
Title is self-explanatory.

Closes #16818.
Closes #18656.

### CODE REVIEW

File: pandas/tests/series/test_analytics.py [hunk 1]
```diff
@@ -2082,6 +2082,18 @@ def test_boundary_datetimelike(self, nselect_method, dtype):
         vals = [min_val + 1, min_val + 2, max_val - 1, max_val, min_val]
         assert_check_nselect_boundary(vals, dtype, nselect_method)
 
+    def test_keep_all_ties(self):
+        # see gh-16818
+        s = Series([10, 9, 8, 7, 7, 7, 7, 6])
+        result = s.nlargest(4, keep='all')
+        expected = Series([10, 9, 8, 7, 7, 7, 7])
+        print(result, expected)
```
Comment (jschendel) [inline pandas/tests/series/test_analytics.py]: `print` statement should be removed

File: doc/source/whatsnew/v0.24.0.txt [hunk 2]
```diff
@@ -24,6 +24,7 @@ Other Enhancements
   <https://pandas-gbq.readthedocs.io/en/latest/changelog.html#changelog-0-5-0>`__.
   (:issue:`21627`)
 - New method :meth:`HDFStore.walk` will recursively walk the group hierarchy of an HDF5 file (:issue:`10932`)
+- :func:`Series` / :func:`DataFrame` methods :func:`nlargest` / :func:`nsmallest` now accept the value 'all' for the `keep` argument. This keeps all ties for the nth largest/smallest value (:issue:`16818`)
```
Comment (jschendel) [inline doc/source/whatsnew/v0.24.0.txt]: I don't think these will link properly if the class is separated from the method (I could be wrong though), and I think you'll need to explicitly write out the variations, e.g. <code>:meth:\`Series.nlargest\`</code>.

**Comment by gfyoung** [comment]
@jschendel : Thanks for comments!  TBH, I didn't really check the PR diff before trying to apply it on `master`.  I'll get to these shortly.

File: pandas/tests/series/test_analytics.py (hunk 1 above)
Comment (gfyoung) [reply pandas/tests/series/test_analytics.py]: Done.

File: doc/source/whatsnew/v0.24.0.txt (hunk 2 above)
Comment (gfyoung) [reply doc/source/whatsnew/v0.24.0.txt]: Done.

**Comment by gfyoung** [comment]
@jreback @jschendel : Comments addressed, and all is passing.  PTAL.

**Comment by jschendel** [comment]
Can you add `'all'` to the list of params for the `Series` and `DataFrame` asv's related to `NSort`:

https://github.com/pandas-dev/pandas/blob/a620e7255ce896c9185275d82172513268b1a719/asv_bench/benchmarks/series_methods.py#L41-L44

https://github.com/pandas-dev/pandas/blob/a620e7255ce896c9185275d82172513268b1a719/asv_bench/benchmarks/frame_methods.py#L501-L504

I don't think this change will have any perf impact, so no need to run the asv's, but might be nice to have benchmarks in place for this.

**Comment by gfyoung** [comment]
@jschendel : Yep, done.

File: pandas/core/frame.py [hunk 3]
```diff
@@ -4645,25 +4658,73 @@ def nsmallest(self, n, columns, keep='first'):
...
+        When using ``keep='last'``, ties are resolved in reverse order:
+
+        >>> df.nsmallest(3, 'a', keep='last')
+           a  b    c
+        0  1  a  1.0
+        5  2  f  9.0
+        4  8  e  4.0
+
+        When using ``keep='all'``, all duplicate items are maintained
+        >>> df.nsmallest(3, 'a', keep='all')
```
Comment (jreback) [inline pandas/core/frame.py]: nitpick, blank line here

File: pandas/core/frame.py [hunk 4]
```diff
@@ -4645,25 +4658,73 @@ def nsmallest(self, n, columns, keep='first'):
...
+           a  b    c
+        0  1  a  1.0
+        5  2  f  9.0
+        4  8  e  4.0
+
+        Attempting to use ``nsmallest`` on non-numeric dtypes will raise a
+        ``TypeError``:
+
+        >>> df.nsmallest(3, 'b')
+        Traceback (most recent call last):
```
Comment (jreback) [inline pandas/core/frame.py]: same here

File: pandas/tests/frame/test_analytics.py [hunk 5]
```diff
@@ -2461,6 +2461,22 @@ def test_n_duplicate_index(self, df_duplicates, n, order):
         expected = df.sort_values(order, ascending=False).head(n)
         tm.assert_frame_equal(result, expected)
 
+    def test_keep_all_ties(self):
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: can you name: test_duplicate_keep_all_ties

side issue: I think make sense to split out duplicate tests to separate file (can you make an issue / PR if you feel!)

File: pandas/tests/series/test_analytics.py [hunk 6]
```diff
@@ -2082,6 +2082,17 @@ def test_boundary_datetimelike(self, nselect_method, dtype):
         vals = [min_val + 1, min_val + 2, max_val - 1, max_val, min_val]
         assert_check_nselect_boundary(vals, dtype, nselect_method)
 
+    def test_keep_all_ties(self):
```
Comment (jreback) [inline pandas/tests/series/test_analytics.py]: same naming here

File: doc/source/whatsnew/v0.24.0.txt [hunk 7]
```diff
@@ -24,6 +24,7 @@ Other Enhancements
   <https://pandas-gbq.readthedocs.io/en/latest/changelog.html#changelog-0-5-0>`__.
   (:issue:`21627`)
 - New method :meth:`HDFStore.walk` will recursively walk the group hierarchy of an HDF5 file (:issue:`10932`)
+- :meth:`Series.nlargest`, :meth:`Series.nsmallest`, :meth:`DataFrame.nlargest`, and :meth:`DataFrame.nsmallest` now accept the value ``"all"`` for the ``keep` argument. This keeps all ties for the nth largest/smallest value (:issue:`16818`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you add both issue numbers here?

**Review by jreback — APPROVED** [review]
lgtm. very minor comments. merge on green.

File: doc/source/whatsnew/v0.24.0.txt (hunk 7 above)
Comment (gfyoung) [reply doc/source/whatsnew/v0.24.0.txt]: The second issue number #18656 is an actually the old PR.  I mentioned it for completeness.

File: pandas/tests/series/test_analytics.py (hunk 6 above)
Comment (gfyoung) [reply pandas/tests/series/test_analytics.py]: Done.

File: pandas/tests/frame/test_analytics.py (hunk 5 above)
Comment (gfyoung) [reply pandas/tests/frame/test_analytics.py]: Done.

File: pandas/core/frame.py (hunk 4 above)
Comment (gfyoung) [reply pandas/core/frame.py]: Done.

File: pandas/core/frame.py (hunk 3 above)
Comment (gfyoung) [reply pandas/core/frame.py]: Done.

**Comment by gfyoung** [comment]
All is green, so merging.

**Pull request merged**